### DISCUSSION TITLE
Programatically setting log level based on name of logger

### DISCUSSION THREAD

**Discussion by wimax-grapl** [pytest-dev/pytest#8925] [Q&A]
Hi folks! Got a caplog question.

I have some personal code, which calls stuff like `urllib3`, `requests`, etc.
My personal code specifies multiple `logging.getLogger(...)` with different names, e.g.
`logging.getLogger("coolproject.module_one")`
`logging.getLogger("coolproject.module_two.submodule")`
stuff like that, you get the gist.

When I specify `pytest --log-level=DEBUG`, obviously that triggers debug statements from not just `coolproject`, but also `urllib3` and `requests`. 

Would there be a way for me to specify something like:
- Globally, log level = INFO
- For any logger with `coolproject` in its name, log level = DEBUG
?

I see that caplog supports stuff like `caplog.set_level(logging.CRITICAL, logger="root.baz")`, but that's a concrete logger name.
What I'd like is a `caplog.set_level(logging.DEBUG, logger_name_regex="coolproject.*")`

**Comment by portyu9** [comment]
Hi @wimax-grapl 

You don't need a regex for this. Python logger names are hierarchical.

If all of your project's loggers are named like:

```text
coolproject
coolproject.foo
coolproject.foo.bar
```

set the parent logger to `DEBUG`:

```python
import logging

logging.getLogger("coolproject").setLevel(logging.DEBUG)
```

Its children inherit that level unless they explicitly override it.

For a particular test, pytest's `caplog` fixture can do the same thing temporarily:

```python
def test_something(caplog):
    caplog.set_level(logging.DEBUG, logger="coolproject")

    ...
```

Other libraries such as `urllib3` and `requests` can remain at their normal levels, or you can explicitly set them higher if needed:

```python
logging.getLogger("urllib3").setLevel(logging.INFO)
```

So the normal solution is to use the logger hierarchy rather than filtering individual records by string prefix.

Please mark this answer as accepted if it helped, thank you.
